core/arch/arm/kernel/stmm_sp.c

```diff
@@ -474,23 +526,48 @@ static int sp_svc_set_mem_attr(vaddr_t va, unsigned int nr_pages, uint32_t perm)
 static bool return_helper(bool panic, uint32_t panic_code,
 			  struct thread_svc_regs *svc_regs)
 {
-	if (!panic) {
-		struct ts_session *sess = ts_get_current_session();
-		struct stmm_ctx *spc = to_stmm_ctx(sess->ctx);
-		size_t n = 0;
+	struct ts_session *sess = ts_get_current_session();
+	struct stmm_ctx *spc = to_stmm_ctx(sess->ctx);
 
+	if (panic) {
+		spc->ta_ctx.panicked = true;
+	} else {
+		size_t __maybe_unused n = 0;
+
+#ifdef ARM64
 		/* Save the return values from StMM */
 		for (n = 0; n <= 7; n++)
 			spc->regs.x[n] = *(&svc_regs->x0 + n);
 
 		spc->regs.sp = svc_regs->sp_el0;
 		spc->regs.pc = svc_regs->elr;
 		spc->regs.cpsr = svc_regs->spsr;
+#endif
+#ifdef ARM32
+		spc->regs.r0 = svc_regs->r0;
+		spc->regs.r1 = svc_regs->r1;
+		spc->regs.r2 = svc_regs->r2;
+		spc->regs.r3 = svc_regs->r3;
+		spc->regs.r4 = svc_regs->r4;
+		spc->regs.r5 = svc_regs->r5;
+		spc->regs.r6 = svc_regs->r6;
+		spc->regs.r7 = svc_regs->r7;
+		spc->regs.pc = svc_regs->lr;
+		spc->regs.usr_sp = thread_a32_get_usr_sp();
```

Comment:
This must be a bit late. I expect that we need to call `thread_a32_get_usr_sp()` before we risk doing any RPC.

Reply:
True.
We could save **usr_sp** in the syscall context, even it no all syscalls expect **usr_sp** backup.

Reply:
Hmm, perhaps this is fine after all. While we're still using the same thread `usr_sp` will be saved in the thread context even if we're migrated to another CPU.

If I understand this correctly it's only needed after an invoke of StMM has completed so the stack pointer get a correct value next time.

Reply:
This is needed if invoking non-secure or if Core execute another user space on this thread.